### ISSUE TITLE
Prevent duplicates when adding data

### ISSUE THREAD

**Issue by technimad** [semi-technologies/weaviate#1161]
# Use Case
In our use case we are adding a dataset of articles to Weaviate. These articles have properties which are unique to each article, like title and abstract, and they have some properties which are shared over multiple articles, like author or publisher. These shared properties are modeled in our schema as individual classes.

Let me give an example to illustrate the issue we run into:
| ID | Title | Author |
|---|---|---|
| 1 | Bugs are cool | John D. |
| 2 | Preferred method of squashing | Jane Doe |
| 3 | Small animals and computers | John D. |
| 4 | Using machines to evict bugs | Marie C. |

_Id column exists only for easy referral to the table in the following text._

# Issue
During import we add an author with name "John D." and use the UUID returned when we add the article with ID 1 to link to this author. We do the same for the next article and so on.

This results in **4 separate** _things_ of class author in Weaviate of which **2 share** the **exact same values** for all their properties (name = "John D."). 
**Off course we want to point the author of article 1 and 3 both to the same _thing_.** 

# Possible current solutions
This is quite a challenge with the current Weaviate API. We've looked at the following:
- Create a hashing function which creates a predictable UUID based on the objects field contents and use that to also create the UUID of author objects, so article 1 and 3 will both point to the same UUID. Still the object has to be created once, so some state has to be kept by the importer,
- Create a key value table during import which consists of (a hash of the properties of) the author and UUID generated by Weaviate. Before creating a new thing, consult the key value table to check if the thing is already created and get the UUID form there or create the thing.
- Before creating a new thing in Weaviate always first query for the existence of a thing with the same properties.

All options have there own drawbacks, but most of all, it creates a hurdle to get data into Weaviate.

# Improvement proposal
A change which would really help in this case is the following:

A POST to ```/v1/{semanticKind}``` should not create a new _thing_ if a _thing_ with the exact same properties already exists. It should return the beacon to the existing _thing_ instead.

I do see that this behaviour is not always the expected or intended behaviour. Switching between the current and proposed behaviour could be done based on:
- query parameter. i.e. ```/v1/{semanticKind}?checkDuplicate=true``` forces the proposed behaviour
- http header. A defined http header i.e. ```X-WEAVIATE-checkDuplicate: true``` is sent with the POST request to indicate the expected behaviour
- specific endpoint. i.e. ```/v1/{semanticKind}/no-duplicates``` to force the proposed behaviour
- a new field in the posted data object

**Comment by etiennedi** [comment]
Hi @technimad 

Thanks for the very thorough issue description.

Regarding 
> This results in 4 separate things of class author in Weaviate of which 2 share the exact same values for all their properties (name = "John D.").
Off course we want to point the author of article 1 and 3 both to the same thing.

When about to import a potential duplicate, can't you do a simple check for a duplicate:
1. Give me all the authors whose name is exactly "John D."
2. If there is a result, take the UUID of that result
3. If there is no result, proceed with importing as planned and take the UUID of the new item

This way you can reuse existing APIs and you are 100% in control of what makes something a duplicate (e.g. you could compare some fields, but not all, etc.)

**Comment by technimad** [comment]
Yes, we could do that. Yet this would make the import code more complex and less performant. Worst case it doubles the calls made to Weaviate. The use case described above is a simplification of reality. In reality we have more than 10 related _things_ for one article.

It's not that Weaviate is lacking functionality. 
Adding functionality which detects duplicates will remove friction during the on-boarding of new data.

In our use case we will deterministically generate the uuid's based on the content of the author _thing_. And always POST to Weaviate. Which works, but feels like a work around which should not be needed.

**Comment by bobvanluijt** [comment]
@technimad do you set UUID's yourself or do you use the dynamically assigned one?

**Comment by technimad** [comment]
We set the uuids. If we add it and it’s a duplicate, the python lib throws a nice exception.

**Comment by bobvanluijt** [comment]
Aha, but that solves your problem, right? Or am I missing something

**Comment by technimad** [comment]
Yes.
